academy/ibc-dev/ibcmw-develop.md

```diff
@@ -0,0 +1,456 @@
+# Create a custom IBC middleware
+
+When developing a custom IBC application, one of the first things to do is to implement the `IBCModule` interface as we saw [previously](insert-link.com).
+
+The interface could be found [here](https://github.com/cosmos/ibc-go/blob/main/modules/core/05-port/types/module.go).
+
+Likewise when developing IBC middleware, the `Middleware` interface should be implemented. It can be found in the same file as the `IBCModule` and looks as follows.
+
+```go
+type Middleware interface {
+    IBCModule
+    ICS4Wrapper
+}
+```
+
+Note that middleware must implement `IBCModule` to wrap communication from core IBC to underlying application and `ICS4Wrapper` to wrap communication from underlying application to core IBC.
+
+The former we know from the IBC custom app section, the latter is an interface like so:
+
+```go
+// This is implemented by ICS4 and all middleware that are wrapping base application.
+// The base application will call `sendPacket` or `writeAcknowledgement` of the middleware directly above them
+// which will call the next middleware until it reaches the core IBC handler.
+type ICS4Wrapper interface {
+    SendPacket(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.Packet) error
+    WriteAcknowledgement(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.Packet, ack exported.Acknowledgement) error
+    GetAppVersion(ctx sdk.Context, portID, channelID string) (string, bool)
+}
+```
+
+When developing custom middleware, we can implement these interfaces for a new `IBCMiddleware` that has access to its keeper and an underlying application. In the `ibc_middleware.go` file:
+
+```go
+import (
+	...
+	porttypes "github.com/cosmos/ibc-go/v5/modules/core/05-port/types"
+	...
+)
+var _ porttypes.Middleware = &IBCMiddleware{}
+
+type IBCMiddleware struct {
+	app    porttypes.IBCModule
+	keeper keeper.Keeper
```

Comment:
middleware only needs its own keeper if it is stateful. stateless middleware do not need a keeper. They can just hold an underlying app

Reply:
Right, will add that comment